### PR TITLE
[Keyboard] Add Synth Labs Solo keyboard

### PR DESCRIPTION
**Pull request by hongaaronc**
## Description

This PR adds the Synth Labs Solo macropad as a supported keyboard in QMK. This has been tested manually on a prototype PCB.

It includes 3 keymaps:
- A default keymap for usage as a macropad
- A via keymap for usage as a macropad with reassignable keys and rotary encoder behavior through VIA
- A gamepad keymap for usage as a Beatmania IIDX controller, which maps the rotary encoder to an emulated joystick

More information about this macropad can be found here: https://geekhack.org/index.php?topic=115880.0

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/synthlabs/solo/keymaps/default/keymap.c
Comment (drashna) [inline keyboards/synthlabs/solo/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/synthlabs/solo/keymaps/default/keymap.c
Comment (drashna) [inline keyboards/synthlabs/solo/keymaps/default/keymap.c:36]: This isn't the right away to do this.

```suggestion
```

File: keyboards/synthlabs/solo/keymaps/gamepad/keymap.c
Comment (drashna) [inline keyboards/synthlabs/solo/keymaps/gamepad/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/synthlabs/solo/keymaps/via/keymap.c
Comment (drashna) [inline keyboards/synthlabs/solo/keymaps/via/keymap.c:36]:
```suggestion
```

File: keyboards/synthlabs/solo/solo.c
Comment (drashna) [inline keyboards/synthlabs/solo/solo.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/synthlabs/solo/solo.h
Comment (drashna) [inline keyboards/synthlabs/solo/solo.h:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/synthlabs/solo/solo.h
Comment (drashna) [inline keyboards/synthlabs/solo/solo.h:6]:
```suggestion
```

File: keyboards/synthlabs/solo/solo.h
Comment (drashna) [inline keyboards/synthlabs/solo/solo.h:40]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/synthlabs/solo/keymaps/default/keymap.c
Comment (hongaaronc) [reply keyboards/synthlabs/solo/keymaps/default/keymap.c:36]: I elected to give the rotary encoder keycodes a virtual position in the key matrix so that they can be remapped in VIA, which has been successful from my testing.

I haven't really seen this functionality elsewhere in the codebase - but maybe there's a better way to do it that I just missed. Any suggestions?

File: keyboards/synthlabs/solo/keymaps/default/keymap.c
Comment (hongaaronc) [reply keyboards/synthlabs/solo/keymaps/default/keymap.c:1]: Sure thing. Will add.

File: keyboards/synthlabs/solo/solo.h
Comment (hongaaronc) [reply keyboards/synthlabs/solo/solo.h:6]: Makes sense. Will remove as well as [line 40](https://github.com/qmk/qmk_firmware/pull/16103#discussion_r794990869).

File: keyboards/synthlabs/solo/keymaps/default/keymap.c
Comment (hongaaronc) [reply keyboards/synthlabs/solo/keymaps/default/keymap.c:4]: Added.

File: keyboards/synthlabs/solo/keymaps/gamepad/keymap.c
Comment (hongaaronc) [reply keyboards/synthlabs/solo/keymaps/gamepad/keymap.c:4]: Added as of 48aa8c696efe739a80c490b2883d5c324945690e.

File: keyboards/synthlabs/solo/solo.c
Comment (hongaaronc) [reply keyboards/synthlabs/solo/solo.c:1]: Added as of 48aa8c696efe739a80c490b2883d5c324945690e.

File: keyboards/synthlabs/solo/solo.h
Comment (hongaaronc) [reply keyboards/synthlabs/solo/solo.h:4]: Added as of 48aa8c696efe739a80c490b2883d5c324945690e.

File: keyboards/synthlabs/solo/keymaps/via/keymap.c
Comment (hongaaronc) [reply keyboards/synthlabs/solo/keymaps/via/keymap.c:39]: Implemented this way for the same reason as [above comment](https://github.com/qmk/qmk_firmware/pull/16103/files#r794993356).

File: keyboards/synthlabs/solo/keymaps/default/keymap.c
Comment (drashna) [reply keyboards/synthlabs/solo/keymaps/default/keymap.c:39]: The issue with this method is that it only supports basic keycodes.  Not even modded keycodes.  And it doesn't support anything that's not basic. 

If you want to do something like this, then this will work much better: 
```c
static uint8_t encoder_state[] = {0};
static keypos_t encoder_cw[] = ENCODERS_CW_KEY;
static keypos_t encoder_ccw[] = ENCODERS_CCW_KEY;
#endif

void encoder_action_unregister(void) {
#ifdef ENCODERS
    for (int index = 0; index < NUM_OF_ENCODERS; ++index) {
        if (encoder_state[index]) {
            keyevent_t encoder_event = (keyevent_t) {
                .key = encoder_state[index] >> 1 ? encoder_cw[index] : encoder_ccw[index],
                .pressed = false,
                .time = (timer_read() | 1)
            };
            encoder_state[index] = 0;
            action_exec(encoder_event);
        }
    }
}

void encoder_action_register(uint8_t index, bool clockwise) {
    keyevent_t encoder_event = (keyevent_t) {
        .key = clockwise ? encoder_cw[index] : encoder_ccw[index],
        .pressed = true,
        .time = (timer_read() | 1)
    };
    encoder_state[index] = (clockwise ^ 1) | (clockwise << 1);
    action_exec(encoder_event);
}

void matrix_scan_user(void) {
    encoder_action_unregister();
}

bool encoder_update_user(uint8_t index, bool clockwise) {
    encoder_action_register(index, clockwise);
    return false;
}
```

However, I would recommend this only for the via keymap, and add something like the following to the keyboard's c file: 

```c
#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        if (clockwise) {
            tap_code_delay(KC_VOLU, 10);
        } else {
            tap_code_delay(KC_VOLD, 10);
        }
    }
    return true;
}
#endif
```

This way, a default behavior is enabled, but via has some custom handling.

Comment (hongaaronc) [reply keyboards/synthlabs/solo/keymaps/default/keymap.c:39]: Ah I see, thanks for the suggestion! I'll look through this and test it out. Will update soon.

And for having this only in the via keycap, that makes perfect sense to me. Will do.

Comment (hongaaronc) [reply keyboards/synthlabs/solo/keymaps/default/keymap.c]: Updated per @drashna's suggestions in 06e02994f184ebe537df9d5ab0e6eb34fe86bccd and cdad9cdf23e28190886bfd3d625e3776603eff10.

File: keyboards/synthlabs/solo/keymaps/via/keymap.c
Comment (hongaaronc) [reply keyboards/synthlabs/solo/keymaps/via/keymap.c]: Updated per @drashna's suggestions in 06e02994f184ebe537df9d5ab0e6eb34fe86bccd and cdad9cdf23e28190886bfd3d625e3776603eff10.

File: keyboards/synthlabs/solo/solo.c
Comment (hongaaronc) [reply keyboards/synthlabs/solo/solo.c:4]: Whoops, sorry. Missed that one - thanks for the catch. Added header as of ac7f06b761afa713891120800f66418198d1b24d.

**Comment by hongaaronc** [comment]
@drashna All requested changes have been addressed. Could you re-review this when you get a chance?

**Comment by 66f94eae** [comment]
> @drashna All requested changes have been addressed. Could you re-review this when you get a chance?

hi @hongaaronc, I'm in the same as you. After making some changes as he said..................the world has not any news

File: keyboards/synthlabs/solo/config.h
Comment (drashna) [inline keyboards/synthlabs/solo/config.h:26]: Couldn't you use `NO_PIN` for these, instead?

File: keyboards/synthlabs/solo/config.h
Comment (drashna) [inline keyboards/synthlabs/solo/config.h:16]: This should be an integer, as it's converted to one, before usage, anyways.

File: keyboards/synthlabs/solo/info.json
Comment (drashna) [inline keyboards/synthlabs/solo/info.json:8]: You don't need the config.h settings if you have them defined here.

File: keyboards/synthlabs/solo/rules.mk
Comment (drashna) [inline keyboards/synthlabs/solo/rules.mk:15]:
```suggestion
NKRO_ENABLE = yes           # Enable N-Key Rollover
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/synthlabs/solo/config.h
Comment (hongaaronc) [reply keyboards/synthlabs/solo/config.h]: Ah, yep! That makes way more sense - thanks for pointing that out. Changed as of 5c9de2621058c1a50e4fd867e77157331cc6acf5 .

Comment (hongaaronc) [reply keyboards/synthlabs/solo/config.h]: Actually, I ended up removing this declaration entirely as per your other suggestion in https://github.com/qmk/qmk_firmware/pull/16103#discussion_r810872075

Replaced these dummy values with `null` in `info.json` as of e1ab7730d4836b1c4e61193470b54ccba087da63

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/synthlabs/solo/readme.md
Comment (hongaaronc) [reply keyboards/synthlabs/solo/readme.md:3]: This existing URI seems to work fine for me, and the suggested one doesn't seem to be valid. Was this suggestion intentional?

Comment (fauxpark) [reply keyboards/synthlabs/solo/readme.md:3]: It's valid: https://i.imgur.com/uVUtfksh.png

Comment (hongaaronc) [reply keyboards/synthlabs/solo/readme.md:3]: Oh whoops. I guess I must have typed something in wrong earlier when I tested it earlier. Yeah, you're right - it works. Will change.

**Comment by BachPXHE163049** [comment]
HI! I'm new here. Does this keypad have e-Amusement Cloud spoof for Infinitas?

**Comment by hongaaronc** [comment]
> HI! I'm new here. Does this keypad have e-Amusement Cloud spoof for Infinitas?

Hey there @BachPXHE163049, thanks for the interest! Feel free to ask any questions on [the Geekhack thread](https://geekhack.org/index.php?topic=115880.0), the [Synth Labs Solo Discord channel](https://discord.gg/NR2N66FTnY), or directly to me via Discord DM (Nostril#0001). I'd like to keep this GitHub discussion focused on this pull request.

**Review by drashna — APPROVED** [review]

**Comment by hongaaronc** [comment]
Still need a second approval before I can merge. There are no outstanding issues, but let me know if any more come up.

**Review by fauxpark — APPROVED** [review]

**Comment by hongaaronc** [comment]
Thanks for the approvals. Looks like "Merging is blocked" though since it's restricted to authorized users - just realized this now. Could someone merge this for me?

**Comment by drashna** [comment]
Thanks!

**Comment by hongaaronc** [comment]
Thank you! 😁

**Pull request merged**
